### DISCUSSION TITLE
Is set_dofs_velocity used to set the base velocity or the initial velocity?

### DISCUSSION THREAD

**Discussion by codesbykhalil** [Genesis-Embodied-AI/genesis-world#1202] [Q&A]
I noticed that after using this function to set the velocity, the object falls and then bounces back, continuing to move in the opposite direction **without stopping.**

**Comment by duburcqa** [comment]
Can you provide a minimal reproducible example ?

**Reply by codesbykhalil** [reply]
import genesis as gs
import trimesh
import os
import time
import numpy as np

def calculate_triangle_center(obj_file):
    """计算三角形的中心点位置"""
    vertices = []
    with open(obj_file, 'r') as f:
        for line in f:
            if line.startswith('v '):  # 只读取顶点行
                # 分割行并转换为浮点数
                _, x, y, z = line.split()
                vertices.append([float(x),float(z),  float(y)]) #置换y与z的坐标,将块体位置投放到xz平面，用于模拟块体坠落
                if len(vertices) == 3:  # 只需要前三个顶点
                    break
    
    if len(vertices) < 3:
        return (0, 0, 0)  # 如果没有足够的顶点，返回原点
        
    # 计算三个顶点的平均值
    vertices = np.array(vertices)
    center = np.mean(vertices, axis=0)
    return tuple(center)  # 返回(x, y, z)元组

def main():
    # 初始化Genesis
    gs.init(backend=gs.gpu,
            theme = 'light',)  # 使用GPU后端
    
    # 创建场景，使用用户给定的参数
    scene = gs.Scene(
        show_viewer    = True,
        viewer_options = gs.options.ViewerOptions(
            res           = (640, 480),  # 降低分辨率
            camera_pos    = (10, 10, 10),  # 调整相机位置到更高的位置
            camera_lookat = (5, 5, 5),  # 看向物体下落的位置
            camera_fov    = 60,  # 增加视场角以看到更多场景
            max_FPS       = 30,  # 降低帧率
        ),
        vis_options = gs.options.VisOptions(
            show_world_frame = True, # 显示原点坐标系
            world_frame_size = 5, # 坐标系长度(米)
            show_link_frame  = False, # 不显示实体链接坐标系 
            show_cameras     = False, # 不显示相机网格和视锥
            plane_reflection = False, # 开启平面反射
            ambient_light    = (0.1, 0.1, 0.1), # 环境光
           
        ),
        # renderer = gs.renderers.Rasterizer(), # 使用光栅化渲染器
        rigid_options=gs.options.RigidOptions(
            enable_collision=True,  # 启用碰撞
            gravity=(0, 0, 0),  # 重力方向保持不变，因为我们是整个场景旋转
        ),
    )
    frictionless_rigid = gs.materials.Rigid(needs_coup=True, coup_friction=0.5)
    # plane = scene.add_entity(
    #     gs.morphs.Plane(
    #         euler=(0, 90, 90),  # 绕Z轴顺时针旋转90度
    #     ),
    # )
    # 读取temp目录下前100个prism_xxx.obj文件
    temp_dir = "examples/obj/temp/temp"
    obj_files = sorted([f for f in os.listdir(temp_dir) if f.endswith('.obj')])[:5]
    print(f"将渲染以下文件: {obj_files}")
    
    # 保存物体引用，用于后续设置初始速度
    prism_objects = []
    
    for obj_file in obj_files:
        file_path = os.path.join(temp_dir, obj_file)
        # 计算三棱柱体积和质量（可选）
        mesh = trimesh.load(file_path)
        volume = mesh.volume
        mass = volume * 10.0  # 假设密度为10
        
        # 计算物体的中心位置
        center_pos = calculate_triangle_center(file_path)
        print(f"物体 {obj_file} 的中心位置: {center_pos}")
        
        # 添加到场景
        obj = scene.add_entity(
            morph=gs.morphs.Mesh(
                file=file_path,
                pos=center_pos,  # 使用计算出的中心位置
                fixed=False,     # 允许物理模拟
            ),
            surface=gs.surfaces.Default(
                color=(0.0, 0.0, 1.0),
                smooth=True,  # 平滑渲染
            ),
        )
        prism_objects.append(obj)

    v_obj_init = gs.tensor([0.0, 2.0, -9.81, 0, 0, 0], requires_grad=True)

    # 添加一个平面作为地面
    scene.add_entity(
        material=frictionless_rigid,
        morph=gs.morphs.Box(
            pos=(0.0, 0.0, -8),  # 位置
            size=(50.0, 50.0, 0.1),  # 大小
            fixed=True,  # 固定地面
        ),
        surface=gs.surfaces.Rough(
            color=(0.8, 0.8, 0.8),
            roughness=100,  
            metallic=0.0,   
            ior=1.5,
        ),
    )
    

    
    # 添加录屏相机（必须在scene.build()之前）
    cam = scene.add_camera(
        res    = (640, 480),
        pos    = (20, 0,20),
        lookat = (5, 0, 5),
        fov    = 40,
        GUI    = False,
    )

    # 构建场景
    scene.build()

    # 录屏主循环
    cam.start_recording()
    steps = 30 * 10  # 10秒，30帧每秒
    for i in range(steps):
        # 在2秒后添加重力和初始速度
        if i == 30 * 2:  # 2秒后
            for obj in prism_objects:
                obj.set_dofs_velocity(v_obj_init)
            print("2秒后开始添加重力和初始速度")
        
        scene.step()
        cam.render()
    cam.stop_recording(save_to_filename='examples/obj/prism_simulation.mp4', fps=30)
    print('录屏已结束，视频已保存。')

    # 继续交互式运行（可选）
    while True:
        scene.step()

if __name__ == "__main__":
    main() 
  **### Asserts:**
[temp.zip](https://github.com/user-attachments/files/20450987/temp.zip)

**Comment by duburcqa** [comment]
Are you aware that set_dofs_velocity only overwrites the current velocity ? This means that at the end of the next step the velocity will not be zero anymore, and the object will keep accelerating under gravity effect. It will never be still even if you overwrite it at every timestep unless the acceleration is truely zero. There is no way to prevent this except fixing the object.

**Reply by duburcqa** [reply]
We could add an option to fix object wrt world dynamically if necessary.

**Comment by codesbykhalil** [comment]
Do you mean that the velocity at each timestep will always be the one set by set_dofs_velocity and won't change?




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Genesis-Embodied-AI/Genesis"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年5月27日(星期二) 下午2:23
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [Genesis-Embodied-AI/Genesis] Is set_dofs_velocity used to set the base velocity or the initial velocity? (Discussion #1202)





 
Are you aware that set_dofs_velocity only overwrites the current velocity ? This means that at the end of the next step the velocity will not be zero anymore, and the object will keep accelerating under gravity effect. It will never be still even if you overwrite it at every timestep unless the acceleration is truely zero. There is no way to prevent this except fixing the object.

**Reply by duburcqa** [reply]
No I mean the complete opposite. The velocity will never be the one set by set_dofs_velocity (except right after calling it) and will always keep changing. The object will keep moving even if you force set_dofs_velocity after every single step.

**Comment by codesbykhalil** [comment]
Yes, I need that. I need to simulate the object eventually coming to rest. Thank you!




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Genesis-Embodied-AI/Genesis"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年5月27日(星期二) 下午2:24
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [Genesis-Embodied-AI/Genesis] Is set_dofs_velocity used to set the base velocity or the initial velocity? (Discussion #1202)





 
We could add an option to fix object wrt world dynamically if necessary.

**Reply by duburcqa** [reply]
Are you aware that if you set the velocity manually at every timestep, then you are basically not simulating anything for this object, just doing animation ?

**Comment by codesbykhalil** [comment]
Oh, I see what you mean now. But what I'm asking is: if I set the initial velocity in the current timestep, and also set gravity to zero with no other acceleration, then the object should move at a constant velocity, right?




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Genesis-Embodied-AI/Genesis"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年5月27日(星期二) 下午2:53
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [Genesis-Embodied-AI/Genesis] Is set_dofs_velocity used to set the base velocity or the initial velocity? (Discussion #1202)





 
No I mean the complete opposite. The velocity will never be the one set by set_dofs_velocity (except right after calling it) and will always keep changing. The object will keep moving even if you force set_dofs_velocity after every single step.

**Reply by duburcqa** [reply]
Yes, which also means no collision or any kind of external force.

**Reply by duburcqa** [reply]
If not the case, then it probably means that dofs damping is non-zero.

**Comment by codesbykhalil** [comment]
Thank you for the reminder, I understand. I'm not trying to do a real physical simulation — I just need to give the object an initial velocity, throw it out, and let it fall to the ground to create an animation video.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Genesis-Embodied-AI/Genesis"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年5月27日(星期二) 下午2:54
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [Genesis-Embodied-AI/Genesis] Is set_dofs_velocity used to set the base velocity or the initial velocity? (Discussion #1202)





 
Are you aware that if you set the velocity, then you are basically not simulating anything for this object, just doing animation ?

**Reply by duburcqa** [reply]
What you are describing IS physics simulation, unless you script the complete motion trajectory yourself!

**Comment by codesbykhalil** [comment]
I thought that in the Genesis environment, giving an object an initial velocity would be enough to easily generate a video of the object falling, without having to calculate the object's position frame by frame myself. But now I understand what you mean — as long as the object is moving, physical simulation is needed to solve its motion. Thank you for your patient response.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Genesis-Embodied-AI/Genesis"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年5月27日(星期二) 下午3:01
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [Genesis-Embodied-AI/Genesis] Is set_dofs_velocity used to set the base velocity or the initial velocity? (Discussion #1202)





 
What you are describing IS physics simulation, unless you script the complete motion trajectory yourself!

**Reply by duburcqa** [reply]
Yes, and you were right! But this is indeed happening because of physics simulation.

**Comment by duburcqa** [comment]
So at this point you are telling me that you are not observing the object failing and stopping when hitting the ground ?

**Comment by codesbykhalil** [comment]
Yes, that's exactly what I'm talking about. Do you have any suggestions?




4o
















------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Genesis-Embodied-AI/Genesis"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年5月27日(星期二) 下午3:11
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [Genesis-Embodied-AI/Genesis] Is set_dofs_velocity used to set the base velocity or the initial velocity? (Discussion #1202)





 
So at this point you are telling me that you are not observing the object failing and stopping when hitting the ground ?

**Reply by duburcqa** [reply]
Can you share the code, there must be something wrong.
